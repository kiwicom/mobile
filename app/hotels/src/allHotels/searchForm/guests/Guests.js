// @flow

import * as React from 'react';
import { View } from 'react-native';
import {
  TouchableItem,
  Color,
  StyleSheet,
  Icon,
  Text,
} from '@kiwicom/react-native-app-shared';

import GuestsPopup from './GuestsPopup';
import type { RoomConfigurationType } from '../SearchParametersType';

type Props = {|
  guests: RoomConfigurationType,
  onChange: (guests: RoomConfigurationType) => void,
|};

type State = {|
  isPopupOpen: boolean,
|};

export default class Guests extends React.Component<Props, State> {
  state = {
    isPopupOpen: false,
  };

  handlePopupToggle = () =>
    this.setState(prevState => ({
      isPopupOpen: !prevState.isPopupOpen,
    }));

  handleChange = (guests: RoomConfigurationType) => this.props.onChange(guests);

  render = () => {
    const { isPopupOpen } = this.state;
    const { guests } = this.props;
    const guestsCount = guests.adultsCount + guests.children.length;
    const title = `${guestsCount} Guest${guestsCount !== 1 ? 's' : ''}`;
    return (
      <View>
        <TouchableItem onPress={this.handlePopupToggle} activeOpacity={0.6}>
          <View style={buttonStyles.buttonWrapper}>
            <Icon
              name="people"
              size={20}
              color={Color.grey.$600}
              style={buttonStyles.icon}
            />
            <Text>{title}</Text>
          </View>
        </TouchableItem>
        <GuestsPopup
          guests={guests}
          onChange={this.handleChange}
          onClose={this.handlePopupToggle}
          isVisible={isPopupOpen}
        />
      </View>
    );
  };
}

const buttonStyles = StyleSheet.create({
  buttonWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 12,
    elevation: 1, // Android only
    backgroundColor: '#fff',
    android: {
      borderRadius: 3,
      height: 48,
    },
    ios: {
      borderRadius: 0,
      height: 47,
    },
  },
  icon: {
    marginRight: 5,
  },
});
