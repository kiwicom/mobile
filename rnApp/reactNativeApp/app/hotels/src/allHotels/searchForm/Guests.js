// @flow

import * as React from 'react';
import { View, StyleSheet } from 'react-native';
import {
  Popup,
  Button,
  Color,
  NumberControl,
} from '@kiwicom/react-native-app-common';

import type { RoomConfigurationType } from './SearchParametersType';

type Props = {|
  guests: RoomConfigurationType,
  onSave: (guests: RoomConfigurationType) => void,
|};

type State = {|
  guests: RoomConfigurationType,
  popupGuests: RoomConfigurationType,
  isPopupOpen: boolean,
|};

export default class Guests extends React.Component<Props, State> {
  state = {
    guests: this.props.guests,
    popupGuests: this.props.guests,
    isPopupOpen: false,
  };

  componentWillReceiveProps = ({ guests }: Props) =>
    this.setState({ guests, popupGuests: guests });

  handlePopupToggle = () =>
    this.setState(prevState => ({
      isPopupOpen: !prevState.isPopupOpen,
      popupGuests: prevState.guests,
    }));

  handleSave = () => this.props.onSave(this.state.popupGuests);

  handleAdultChange = (adultsCount: number) =>
    this.setState(prevState => ({
      popupGuests: {
        adultsCount,
        children: prevState.popupGuests.children,
      },
    }));

  handleChildrenChange = () => {
    // TODO!
  };

  render = () => {
    const { isPopupOpen, popupGuests, guests } = this.state;
    const guestsCount = guests.adultsCount + guests.children.length;
    const title = `${guestsCount} Guest${guestsCount !== 1 ? 's' : ''}`;
    return (
      <View>
        <Button
          onPress={this.handlePopupToggle}
          title={title}
          styles={buttonStyles}
          icon={{
            name: 'people',
          }}
        />
        <Popup
          onSave={this.handleSave}
          onClose={this.handlePopupToggle}
          isVisible={isPopupOpen}
        >
          <NumberControl
            label="Adult"
            number={popupGuests.adultsCount}
            min={1}
            max={30}
            style={{ marginBottom: 20 }}
            icon="person"
            onChange={this.handleAdultChange}
          />
          <NumberControl
            label="Infant"
            number={popupGuests.children.length}
            min={0}
            max={10}
            icon="child-friendly"
            onChange={this.handleChildrenChange}
          />
        </Popup>
      </View>
    );
  };
}

const buttonStyles = StyleSheet.create({
  button: {
    width: 85,
    backgroundColor: '#fff',
    borderRadius: 0,
    height: 40,
    paddingLeft: 0,
  },
  buttonText: {
    color: Color.grey.$900,
    fontWeight: 'normal',
  },
  icon: {
    paddingLeft: 10,
    paddingRight: 0,
    backgroundColor: '#fff',
    borderTopLeftRadius: 0,
    borderBottomLeftRadius: 0,
  },
});
