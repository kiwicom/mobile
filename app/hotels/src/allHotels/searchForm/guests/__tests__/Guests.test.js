// @flow

import * as React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import { IncrementDecrementButtons } from '@kiwicom/react-native-app-shared';

import Guests from '../Guests';
import type { RoomConfigurationType } from '../../SearchParametersType';

const renderGuests = (onSave: Function) => {
  const guests: RoomConfigurationType = {
    adultsCount: 1,
    children: [],
  };
  return renderer.create(<Guests onChange={onSave} guests={guests} />);
};

const shallowRenderGuests = (onSave: Function) => {
  const renderer = new ShallowRenderer();
  const guests: RoomConfigurationType = {
    adultsCount: 3,
    children: [],
  };
  return renderer.render(<Guests onChange={onSave} guests={guests} />);
};

describe('Guests', () => {
  it('Render all inputs', async () => {
    const renderer = shallowRenderGuests(jest.fn());
    expect(renderer).toMatchSnapshot();
  });

  it('count should be changed after increments and onSave call', async () => {
    expect.assertions(1);

    const onSave = jest.fn();
    const testRenderer = renderGuests(onSave);
    const testInstance = testRenderer.root;

    // Increment adultsCount by 2
    const buttons = testInstance.findAllByType(IncrementDecrementButtons)[0];
    await buttons.props.onIncrement();
    await buttons.props.onIncrement();

    // Save guests count
    await testInstance
      .findByProps({ testID: 'guestsPopupSaveButton' })
      .props.onSave();

    expect(onSave).toBeCalledWith({ adultsCount: 3, children: [] });
  });
});
