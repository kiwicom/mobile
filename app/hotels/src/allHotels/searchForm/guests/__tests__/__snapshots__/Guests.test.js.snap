// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Guests Render all inputs 1`] = `
<View>
  <TouchableItem
    activeOpacity={0.6}
    borderless={false}
    onPress={[Function]}
    pressColor="rgba(0, 0, 0, .32)"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 0,
          "elevation": 1,
          "flexDirection": "row",
          "height": 47,
          "padding": 12,
        }
      }
    >
      <Icon
        color="#757575"
        name="people"
        size={20}
        style={
          Object {
            "marginRight": 5,
          }
        }
      />
      <Text>
        3 Guests
      </Text>
    </View>
  </TouchableItem>
  <GuestsPopup
    guests={
      Object {
        "adultsCount": 3,
        "children": Array [],
      }
    }
    isVisible={false}
    onChange={[Function]}
    onClose={[Function]}
  />
</View>
`;
